src/eexprs.adoc

```diff
@@ -0,0 +1,100 @@
+[[sec:eexprs]]
+== Encoding Expressions
+
+Understanding macro signatures, we can now discuss how macros are leveraged to
+encode data.  The syntax for macro invocation is called an _encoding expression_
+or E-expression.  When the Ion parser encounters an E-expression, it automatically
+replaces it with the values produced by the corresponding macro transformation
+function.  The inputs to that transformation are determined by the arguments
+within the E-expression.
+
+IMPORTANT: This chapter details the syntax of E-expressions in Ion text format;
+the corresponding <<bin:eexp,Ion binary encoding>> enforces the
+equivalent constraints.
+
+In Ion Text, E-expressions look similar to S-expressions but are opened by `(:`
+and a reference to the macro that expands the expression.
+Text-encoded E-expressions have one of several forms that differ in how they
+reference the macro to be invoked:
+
+* `(:__macro-name__ …)` lookup by unambiguous name in system or local macro tables.
+* `(:__address__ …)` lookup by address in current macro table (not the system table).
+* `(:__module-name__:__macro-name__ …)` lookup by name in an installed module.
+* `(:__module-name__:__address__ …)` lookup by address in an installed module.
+
+// TODO link or write more precise resolution rules.
+
+NOTE: The parenthesis, colon, and macro reference are a single syntactic token,
+allowing no white space.
+The names are neither string nor symbol tokens, and thus may not use quotes or
+escapes or `$_uint_` symbol-table addresses.
+This reflects the idea that macros behave like new syntactic forms, with this
+entire character sequence determining the syntax that follows inside the
+expression.
+
+Following the opening macro reference and whitespace, E-expressions follow
+S-expression tokenization and whitespace rules.
+The remaining elements are arguments supplying inputs to the macro,
+each one either _individual_ or _grouped_.
+The syntax that can appear in each argument position is constrained by the
+macro’s signature.
+
+The number of argument elements (that is, the invocation’s actual arity)
+must be equal to or greater than the macro’s minimum arity, and at most its
+maximum arity, when one exists.
+In other words, an E-expression must contain one element for each required
+parameter, followed by optional elements for the remaining optional parameters.
+
+// TODO links for arity, optional parameters
+// TODO base type? base shape?  base form?  encoding?
+
+The syntax of each argument is defined by the associated parameter’s cardinality
+and type.
+The cardinality determines whether an argument group can be used to collect a
+series of individual arguments, and the type determines the syntax of an
+individual argument, grouped or not.
+
+The parameter type constrains the syntax of an individual argument as follows:
+
+* For tagged types, an individual argument must be a single E-expression or any
+  datum (which may contain nested E-expressions).
```

Comment:
We've not defined 'datum', and it might not be the best term. (Other than this line it's only used in the grammar: `(**literal** _datum_)`.  

I'm using it here to mean "syntax tree" as the semantic model defines it, but that term hasn't been incorporated into this document.

"Value" is not correct; since these are syntax trees and we use "value" to denote the results of expanding syntax trees.

Another way to write this might be:

> For tagged types, an individual argument must be a single syntax tree, which may be an E-expression and/or contain nested E-expressions.

But ATM that doesn't seem better since we don't currently talk in detail about the expansion process.  (Another gap in the document!)